### DISCUSSION TITLE
EmulatedPushButtons - comiplation problem and question

### DISCUSSION THREAD

**Discussion by aLAN-LDZ** [HomeSpan/HomeSpan#17] [Q&A] [answered]
Hello
I have a problem compiling the Emulated Push Buttons example.

`DEV_Identify.h:22:13: error: 'LED_BUILTIN' was not declared in this scope
     pinMode(LED_BUILTIN,OUTPUT);                      // make sure built-in LED is set for output
             ^
sketch\DEV_Identify.h: In member function 'virtual boolean DEV_Identify::update()':
DEV_Identify.h:28:20: error: 'LED_BUILTIN' was not declared in this scope
       digitalWrite(LED_BUILTIN,LOW);
`
I understand that with this example I am able to execute any piece of code. After "pressing" the button. I plan to control a cat feeder. Where in the emulatedPushButtons example can I add my code that will control the servo mechanism?

Can i replace this `new DEV_Blinker(16,3);`?

**Comment by HomeSpan** [comment]
You should be able to compile any of the examples, but I was basing that on the assumption that LED_BUILTIN is always defined in the Arduino ecosystem.  Have you modified it all all, or are you just trying to compile the example as is?   If LED_BUILTIN is not defined, none of the other examples should have been able to compile since LED_BUILTIN is also used in the HomeSpan library itself.

**Comment by aLAN-LDZ** [comment]
I did not modify the example and did not define Led_Builtin anywhere.

I just wanted to compile it as is for the test.

I have to rewrite the program so that it executes a fragment of the given code after "pressing"
button on the phone

Could you tell me what I need to change?

**Comment by HomeSpan** [comment]
Yes, the code you would need to add should be placed in the update() method defined (in the case of the Emulated Button Example) in DEV_Blinker.h.  Example 5 takes you through all the steps and commands to use within the update() method.

Also, in terms of compiling, which example did you originally use to test the wifi and pairing?  If one example compiles, all examples *should* compile.

**Comment by aLAN-LDZ** [comment]
I tested the wifi connection and pairing on the example "simple light bulb" and "advanced ceiling fan" in these examples there is no "ledbuiltin" probably for that i manage to compile them. I'm a novice programmer and it's all a bit confused. It is a pity that there is no example of the simplest on / off switch, it would be the easiest way to convert it :(

**Comment by HomeSpan** [comment, marked as answer]
You should definitely start with Example 5.  It's labeled as a lightbulb, but it's just on/off and can be used to trigger anything else.  You can always update it after to convert the Service from LightBulb to Switch.   You would drop your servo code into the update() method, replacing the line that says:

```
digitalWrite(ledPin,power->getNewVal());
```
 with whatever your servo requires.

Within the Arduino IDE, what Board did you select when compiling?  There are many types of ESP32s - I'll see if I can debug the LED_BUILTIN problem.

**Reply by aLAN-LDZ** [reply]
This should be enough for me to continue playing :) maybe the cat will be happy. I am using ESP32 DEV Module for compilation

**Comment by aLAN-LDZ** [comment]
thanks for the hint :) everything works!

**Comment by HomeSpan** [comment]
Glad everything is working. 

Note I just released v1.1.3 that *should* fix a number of the issues discussed, including the problems with LED_BUILTIN not being available for all boards (which prevented compiling); problems with the WiFi initialization logic; and problems with interpreting special characters in a WiFi password when set with the HomeSpan Access Point Web Page.
